File: ClientRuntimes/Ruby/ms-rest-azure/lib/ms_rest_azure/azure_service_client.rb
# encoding: utf-8
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.

module MsRestAzure
  #
  # Class which represents a point of access to the REST API.
  #
  class AzureServiceClient < MsRest::ServiceClient

    # @return [Integer] execution interval for long running operations.
    attr_accessor :long_running_operation_retry_timeout

    # @return [String] api version of the Azure in string format.
    attr_accessor :api_version

    #
    # Retrieves the result of 'POST','DELETE','PUT' or 'PATCH' operation. Performs polling of required.
    # @param azure_response [MsRestAzure::AzureOperationResponse] response from Azure service.
    # @param custom_deserialization_block [Proc] custom logic for response deserialization.
    #
    # @return [MsRest::HttpOperationResponse] the response.
    def get_long_running_operation_result(azure_response, custom_deserialization_block)
      check_for_status_code_failure(azure_response)

      http_method = azure_response.request.method

      polling_state = PollingState.new(azure_response, @long_running_operation_retry_timeout)
      request = azure_response.request

      if !AsyncOperationStatus.is_terminal_status(polling_state.status)
        task = Concurrent::TimerTask.new do
          begin
            if !polling_state.azure_async_operation_header_link.nil?
              update_state_from_azure_async_operation_header(polling_state.get_request(headers: request.headers, base_uri: request.base_uri), polling_state)
            elsif !polling_state.location_header_link.nil?
              update_state_from_location_header(polling_state.get_request(headers: request.headers, base_uri: request.base_uri), polling_state, custom_deserialization_block)
            elsif http_method === :put
              get_request = MsRest::HttpOperationRequest.new(request.base_uri, request.build_path.to_s, :get, query_params: request.query_params)
              update_state_from_get_resource_operation(get_request, polling_state, custom_deserialization_block)
            else
              task.shutdown
              fail AzureOperationError, 'Location header is missing from long running operation'
            end

            if AsyncOperationStatus.is_terminal_status(polling_state.status)
              task.shutdown
            end
          rescue Exception => e
            task.shutdown
            e
          end
        end

        polling_delay = polling_state.get_delay
        polling_delay = 0.1 if polling_delay.nil? || polling_delay == 0

        task.execution_interval = polling_delay
        task.execute
        task.wait_for_termination

        polling_error = task.value
        fail polling_error if polling_error.is_a?(Exception)
      end

      if (http_method === :put || http_method === :patch) && AsyncOperationStatus.is_successful_status(polling_state.status) && polling_state.resource.nil?
        get_request = MsRest::HttpOperationRequest.new(request.base_uri, request.build_path.to_s, :get, query_params: request.query_params)
        update_state_from_get_resource_operation(get_request, polling_state, custom_deserialization_block)
      end

      if AsyncOperationStatus.is_failed_status(polling_state.status)
        fail polling_state.get_operation_error
      end

      polling_state.get_operation_response
    end

    # @TODO Update name of the method to get_put_or_patch_operation_result when introducing breaking change / updating major version of gem
    # Retrieves the result of 'PUT' or 'PATCH' operation. Performs polling of required.
    # @param azure_response [MsRestAzure::AzureOperationResponse] response from Azure service.
    # @param custom_deserialization_block [Proc] custom logic for response deserialization.
    #
    # @return [MsRest::HttpOperationResponse] the response.
    def get_put_operation_result(azure_response, custom_deserialization_block)
      get_long_running_operation_result(azure_response, custom_deserialization_block)
    end

    #
    # Retrieves the result of 'POST' or 'DELETE' operations. Performs polling of required.
    # @param azure_response [MsRestAzure::AzureOperationResponse] response from Azure service.
    # @param custom_deserialization_block [Proc] custom logic for response deserialization.
    #
    # @return [MsRest::HttpOperationResponse] the response.
    def get_post_or_delete_operation_result(azure_response, custom_deserialization_block)
      get_long_running_operation_result(azure_response, custom_deserialization_block)
    end

    #
    # Verifies for unexpected polling status code
    # @param azure_response [MsRestAzure::AzureOperationResponse] response from Azure service.
    def check_for_status_code_failure(azure_response)
      fail MsRest::ValidationError, 'Azure response cannot be nil' if azure_response.nil?
      fail MsRest::ValidationError, 'Azure response cannot have empty response object' if azure_response.response.nil?
      fail MsRest::ValidationError, 'Azure response cannot have empty request object' if azure_response.request.nil?

      status_code = azure_response.response.status
      http_method = azure_response.request.method

      fail AzureOperationError, "Unexpected polling status code from long running operation #{status_code}" unless status_code === 200 || status_code === 202 ||
          (status_code === 201 && http_method === :put) ||
          (status_code === 204 && (http_method === :delete || http_method === :post))
    end

    #
    # Updates polling state based on location header for PUT HTTP requests.
    # @param request [MsRest::HttpOperationRequest] The url retrieve data from.
    # @param polling_state [MsRestAzure::PollingState] polling state to update.
    # @param custom_deserialization_block [Proc] custom deserialization method for parsing response.
    #
    def update_state_from_get_resource_operation(request, polling_state, custom_deserialization_block)
      result = get_async_with_custom_deserialization(request, custom_deserialization_block)

      fail AzureOperationError, 'The response from long running operation does not contain a body' if result.response.body.nil? || result.response.body.empty?

      if result.body.respond_to?(:properties) && result.body.properties.respond_to?(:provisioning_state) && !result.body.properties.provisioning_state.nil?
        polling_state.status = result.body.properties.provisioning_state
      else
        polling_state.status = AsyncOperationStatus::SUCCESS_STATUS
      end

      error_data = CloudErrorData.new
      error_data.code = polling_state.status
      error_data.message = "Long running operation failed with status #{polling_state.status}"

      polling_state.error_data = error_data
      polling_state.update_response(result.response)
      polling_state.request = result.request
      polling_state.resource = result.body
    end

    #
    # Updates polling state based on location header for HTTP requests.
    # @param request [MsRest::HttpOperationRequest] The url retrieve data from.
    # @param polling_state [MsRestAzure::PollingState] polling state to update.
    # @param custom_deserialization_block [Proc] custom deserialization method for parsing response.
    def update_state_from_location_header(request, polling_state, custom_deserialization_block)
      result = get_async_with_custom_deserialization(request, custom_deserialization_block)

      polling_state.update_response(result.response)
      polling_state.request = result.request
      status_code = result.response.status
      http_method = request.method

      if status_code === 202
        polling_state.status = AsyncOperationStatus::IN_PROGRESS_STATUS
      elsif status_code === 200 || (status_code === 201 && http_method === :put) ||
          (status_code === 204 && (http_method === :delete || http_method === :post || http_method === :get))
        polling_state.status = AsyncOperationStatus::SUCCESS_STATUS

        error_data = CloudErrorData.new
        error_data.code = polling_state.status
        error_data.message = "Long running operation failed with status #{polling_state.status}"

        polling_state.error_data = error_data
        polling_state.resource = result.body
      else
        fail AzureOperationError, "The response from long running operation does not have a valid status code. Method: #{http_method}, Status Code: #{status_code}"
      end
    end

    #
    # Updates polling state from Azure async operation header.
    # @param polling_state [MsRestAzure::PollingState] polling state.
    def update_state_from_azure_async_operation_header(request, polling_state)
      result = get_async_with_async_operation_deserialization(request)

      fail AzureOperationError, 'The response from long running operation does not contain a body' if result.body.nil? || result.body.status.nil?

      polling_state.status = result.body.status
      polling_state.error_data = result.body.error
      polling_state.response = result.response
      polling_state.request = result.request
      polling_state.resource = nil
    end

    #
    # Retrieves data by given URL.
    # @param request [MsRest::HttpOperationRequest] the URL.
    # @param custom_deserialization_block [Proc] function to perform deserialization of the HTTP response.
    #
    # @return [MsRest::HttpOperationResponse] the response.
    def get_async_with_custom_deserialization(request, custom_deserialization_block)
      result = get_async_common(request)

      if !result.body.nil? && !custom_deserialization_block.nil?
        begin
          result.body = custom_deserialization_block.call(result.body)
        rescue Exception => e
          fail MsRest::DeserializationError.new("Error occured in deserializing the response", e.message, e.backtrace, http_response.body)
        end
      end

      result
    end

    #
    # Retrieves data by given URL.
    # @param request [MsRest::HttpOperationRequest] the URL.
    #
    # @return [MsRest::HttpOperationResponse] the response.
    def get_async_with_async_operation_deserialization(request)
      result = get_async_common(request)

      result.body = AsyncOperationStatus.deserialize_object(result.body)
      result
    end

    #
    # Retrieves data by given URL.
    # @param request [MsRest::HttpOperationRequest] the URL.
    #
    # @return [MsRest::HttpOperationResponse] the response.
    def get_async_common(request)
      fail ValidationError, 'Request cannot be nil' if request.nil?

      request.middlewares = [[MsRest::RetryPolicyMiddleware, times: 3, retry: 0.02], [:cookie_jar]]
      request.headers.merge!({'x-ms-client-request-id' => SecureRandom.uuid, 'Content-Type' => 'application/json'})

      # Send Request
      http_response = request.run_promise do |req|
        @credentials.sign_request(req) unless @credentials.nil?
      end.execute.value!

      status_code = http_response.status

      if status_code != 200 && status_code != 201 && status_code != 202 && status_code != 204
        json_error_data = JSON.load(http_response.body)
        error_data = CloudErrorData.deserialize_object(json_error_data)

        fail AzureOperationError.new request, http_response, error_data, "Long running operation failed with status #{status_code}"
      end

      result = MsRest::HttpOperationResponse.new(request, http_response, http_response.body)

      begin
        result.body = JSON.load(http_response.body) unless http_response.body.to_s.empty?
      rescue Exception => e
        fail MsRest::DeserializationError.new("Error occured in deserializing the response", e.message, e.backtrace, result)
      end

      result
    end
  end

end

File: src/client/NodeJS/ms-rest-azure/lib/azureServiceClient.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information. 
'use strict';

var util = require('util');
var async = require('async');
var msRest = require('ms-rest');
var PollingState = require('./pollingState');
var LroStates = require('./constants').LongRunningOperationStates;
var WebResource = msRest.WebResource;

/**
 * @class
 * Initializes a new instance of the AzureServiceClient class.
 * @constructor
 * @param {object} credentials - ApplicationTokenCredentials or 
 * UserTokenCredentials object used for authentication.  
 * 
 * @param {object} options - The parameter options used by ServiceClient
 *
 * @param {string} [options.acceptLanguage] - Gets or sets the preferred language for the response. 
 * Default value is: 'en-US'.
 *  
 * @param {boolean} [options.generateClientRequestId] - When set to true a unique x-ms-client-request-id value 
 * is generated and included in each request. Default is true.
 * 
 * @param {number} [options.longRunningOperationRetryTimeout] - Gets or sets the retry timeout in seconds for 
 * Long Running Operations. Default value is 30.
 *
 */
function AzureServiceClient(credentials, options) {
  if (!credentials) {
    throw new Error('Azure clients require credentials.');
  }
  
  AzureServiceClient['super_'].call(this, credentials, options);
  
  this.acceptLanguage = 'en-US';
  this.generateClientRequestId = true;
  this.longRunningOperationRetryTimeout = 30;

  if (!options) options = {};
  
  if (options.acceptLanguage !== null && options.acceptLanguage !== undefined) {
    this.acceptLanguage = options.acceptLanguage;
  }

  if (options.generateClientRequestId !== null && options.generateClientRequestId !== undefined) {
    this.generateClientRequestId = options.generateClientRequestId;
  }

  if (options.longRunningOperationRetryTimeout !== null && options.longRunningOperationRetryTimeout !== undefined) {
    this.longRunningOperationRetryTimeout = options.longRunningOperationRetryTimeout;
  }
}

util.inherits(AzureServiceClient, msRest.ServiceClient);

/**
 * Poll Azure long running PUT or PATCH operation. (Deprecated, new version of the code-gen will generate code to call getLongRunningOperationResult)
 * @param {object} [resultOfInitialRequest] - Response of the initial request for the long running operation.
 * @param {object} [options]
 * @param {object} [options.customHeaders] headers that will be added to request
 */
AzureServiceClient.prototype.getPutOrPatchOperationResult = function (resultOfInitialRequest, options, callback) {
  return this.getLongRunningOperationResult(resultOfInitialRequest, options, callback);
};

/**
 * Poll Azure long running POST or DELETE operations. (Deprecated, new version of the code-gen will generate code to call getLongRunningOperationResult)
 * @param {object} [resultOfInitialRequest] - result of the initial request.
 * @param {object} [options]
 * @param {object} [options.customHeaders] headers that will be added to request
 */
AzureServiceClient.prototype.getPostOrDeleteOperationResult = function (resultOfInitialRequest, options, callback) {
  return this.getLongRunningOperationResult(resultOfInitialRequest, options, callback);
};

/**
 * Poll Azure long running PUT, PATCH, POST or DELETE operations.
 * @param {object} [resultOfInitialRequest] - result of the initial request.
 * @param {object} [options]
 * @param {object} [options.customHeaders] headers that will be added to request
 */
AzureServiceClient.prototype.getLongRunningOperationResult = function (resultOfInitialRequest, options, callback) {
  var self = this;
  
  if (!callback && typeof options === 'function') {
    callback = options;
    options = null;
  }
  if (!callback) {
    throw new Error('Missing callback');
  }

  if (!resultOfInitialRequest) {
    return callback(new Error('Missing resultOfInitialRequest parameter'));
  }
  
  if (!resultOfInitialRequest.response) {
    return callback(new Error('Missing resultOfInitialRequest.response'));
  }
  
  if (!resultOfInitialRequest.request) {
    return callback(new Error('Missing resultOfInitialRequest.request'));
  }
  
  if (!resultOfInitialRequest.request.method) {
    return callback(new Error('Missing resultOfInitialRequest.request.method'));
  }

  var initialRequestMethod = resultOfInitialRequest.request.method;
  
  if (this._checkResponseStatusCodeFailed(resultOfInitialRequest)) {
    return callback(new Error(util.format('Unexpected polling status code from long running operation \'%s\' for method \'%s\'',
      resultOfInitialRequest.response.statusCode,
	  initialRequestMethod)));
  }
  
  var pollingState = null;
  
  try {
    pollingState = new PollingState(resultOfInitialRequest, this.longRunningOperationRetryTimeout);
  } catch (error) {
    callback(error);
  }
  var resourceUrl = resultOfInitialRequest.request.url;
  this._options = options;

  async.whilst(
    function() {
      var finished = [LroStates.Succeeded, LroStates.Failed, LroStates.Canceled].some(function(e) {
        return e === pollingState.status;
      });
      return !finished;
    },
    function (callback) {
      setTimeout(function() {
        if (pollingState.azureAsyncOperationHeaderLink) {
          self._updateStateFromAzureAsyncOperationHeader(pollingState, true, function(err) {
            return callback(err);
          });
        } else if (pollingState.locationHeaderLink) {
          self._updateStateFromLocationHeader(initialRequestMethod, pollingState, function(err) {
            return callback(err);
          });
        } else if (initialRequestMethod === 'PUT') {
          self._updateStateFromGetResourceOperation(resourceUrl, pollingState, function(err) {
            return callback(err);
          });
        } else {
          return callback(new Error('Location header is missing from long running operation.'));
        }
      }, pollingState.getTimeout());
    },
    //when done
    function (err) {
      if (pollingState.status === LroStates.Succeeded) {
        if ((pollingState.azureAsyncOperationHeaderLink || !pollingState.resource) &&
			(initialRequestMethod === 'PUT' || initialRequestMethod === 'PATCH')) {
          self._updateStateFromGetResourceOperation(resourceUrl, pollingState, function(err) {
            return callback(err, pollingState.getOperationResponse());
          });
        } else {
          return callback(null, pollingState.getOperationResponse());
        }
      } else {
        return callback(pollingState.getCloudError(err));
      }
    });
};

AzureServiceClient.prototype._checkResponseStatusCodeFailed = function (initialRequest) {
  var statusCode = initialRequest.response.statusCode;
  var method = initialRequest.request.method;
  if (statusCode === 200 || statusCode === 202 ||
    (statusCode === 201 && method === 'PUT') ||
    (statusCode === 204 && (method === 'DELETE' || method === 'POST'))) {
    return false;
  } else {
	  return true;
  }
};


/**
 * Retrieve operation status by polling from 'azure-asyncoperation' header.
 * @param {object} [pollingState] - The object to persist current operation state.
 * @param {boolean} [inPostOrDelete] - Invoked by Post Or Delete operation.
 */
AzureServiceClient.prototype._updateStateFromAzureAsyncOperationHeader = function (pollingState, inPostOrDelete, callback) {
  this._getStatus(pollingState.azureAsyncOperationHeaderLink, function (err, result) {
    if (err) return callback(err);
    
    if (!result.body || !result.body.status) {
      return callback(new Error('The response from long running operation does not contain a body.'));
    }
    
    pollingState.status = result.body.status;
    pollingState.error = result.body.error;
    pollingState.response = result.response;
    pollingState.request = result.request;
    pollingState.resource = null;
    if (inPostOrDelete) {
      pollingState.resource = result.body;
    }
    callback(null);
  });
};

/**
 * Retrieve PUT operation status by polling from 'location' header.
 * @param {object} [pollingState] - The object to persist current operation state.
 */
AzureServiceClient.prototype._updateStateFromLocationHeader = function (method, pollingState, callback) {
  this._getStatus(pollingState.locationHeaderLink, function(err, result) {
    if (err) return callback(err);
    
    pollingState.updateResponse(result.response);
    pollingState.request = result.request;
    
    var statusCode = result.response.statusCode;
    if (statusCode === 202) {
      pollingState.status = LroStates.InProgress;
    } else if (statusCode === 200 ||
      (statusCode === 201 && method === 'PUT') ||
      (statusCode === 204 && (method === 'DELETE' || method === 'POST'))) {

      pollingState.status = LroStates.Succeeded;

      pollingState.error = {
        code: pollingState.Status,
        message: util.format('Long running operation failed with status \'%s\'.', pollingState.status)
      };
      pollingState.resource = result.body;
    } else {
	  return callback(new Error('The response from long running operation does not have a valid status code.'));
	}
    callback(null);
  });
};

/**
 * Polling for resource status.
 * @param {function} [resourceUrl] - The url of resource.
 * @param {object} [pollingState] - The object to persist current operation state.
 */
AzureServiceClient.prototype._updateStateFromGetResourceOperation = function (resourceUrl, pollingState, callback) {
  this._getStatus(resourceUrl, function (err, result) {
    if (err) return callback(err);
    if (!result.body) {
      return callback(new Error('The response from long running operation does not contain a body.'));
    }
    
    if (result.body.properties && result.body.properties.provisioningState) {
      pollingState.status = result.body.properties.provisioningState;
    } else {
      pollingState.status = LroStates.Succeeded;
    }
    
    //we might not throw an error, but initialize here just in case.
    pollingState.error = {
      code: pollingState.status,
      message: util.format('Long running operation failed with status \'%s\'.', pollingState.status)
    };
    
    pollingState.updateResponse(result.response);
    pollingState.request = result.request;
    pollingState.resource = result.body;
    
    //nothing to return, the 'pollingState' has all the info we care.
    callback(null);
  });
};

/**
 * Retrieve operation status by querying the operation URL.
 * @param {string} [operationUrl] - URL used to poll operation result.
 */
AzureServiceClient.prototype._getStatus = function (operationUrl, callback) {
  var self = this;
  if (!operationUrl) {
    return callback(new Error('operationUrl cannot be null.'));
  }
  
  // Construct URL
  var requestUrl = operationUrl.replace(' ', '%20');
  
  // Create HTTP transport objects
  var httpRequest = new WebResource();
  httpRequest.method = 'GET';
  httpRequest.headers = {};
  httpRequest.url = requestUrl;
  if(this._options) {
    for (var headerName in this._options['customHeaders']) {
      if (this._options['customHeaders'].hasOwnProperty(headerName)) {
        httpRequest.headers[headerName] = this._options['customHeaders'][headerName];
      }
    }
  }
  // Send Request
  return self.pipeline(httpRequest, function (err, response, responseBody) {
    if (err) {
      return callback(err);
    }
    var statusCode = response.statusCode;
    if (statusCode !== 200 && statusCode !== 201 && statusCode !== 202 && statusCode !== 204) {
      var error = new Error(util.format('Invalid status code with response body "%s" occurred ' + 
        'when polling for operation status.', responseBody));
      error.statusCode = response.statusCode;
      error.request = msRest.stripRequest(httpRequest);
      error.response = msRest.stripResponse(response);
      if (responseBody === '') responseBody = null;
      try {
        error.body = JSON.parse(responseBody);

      } catch (badResponse) {
        error.message += util.format(' Could not deserialize error response body - "%s" -.', responseBody);
        error.body = responseBody;
      }

      return callback(error);
    }
    // Create Result
    var result = new msRest.HttpOperationResponse();
    result.request = httpRequest;
    result.response = response;
    if (responseBody === '') responseBody = null;
    if (statusCode === 200 || statusCode === 201 || statusCode === 202) {
      try {
        result.body = JSON.parse(responseBody);
      } catch (deserializationError) {
        var parseError = new Error(util.format('Error "%s" occurred in deserializing the response body - "%s" -' + 
        ' when polling for operation status.', deserializationError, responseBody));
        parseError.request = msRest.stripRequest(httpRequest);
        parseError.response = msRest.stripResponse(response);
        parseError.body = responseBody;
        return callback(parseError);
      }
    }
    return callback(null, result);
  });
};

module.exports = AzureServiceClient;